File: problems/sorting/1268-search-suggestions-system.js
/**
 * #1268 搜索推荐系统
 * 标签：排序、字符串
 * 状态：已完成
 *
 * 时间复杂度：O(p × n × s)
 * 空间复杂度：O(p × 3)
 */

/**
 * @param {string[]} products
 * @param {string} searchWord
 * @return {string[][]}
 */
var suggestedProducts = function (products, searchWord) {
  products.sort((a, b) => {
    for (let i = 0; i < Math.min(a.length, b.length); i++) {
      if (a.charCodeAt(i) !== b.charCodeAt(i)) {
        return a.charCodeAt(i) - b.charCodeAt(i)
      }
    }
  })

  let result = []
  for (let i = 0; i < searchWord.length; i++) {
    let str = searchWord.substring(0, i + 1)

    result.push(
      products
        .filter((item) => item.indexOf(str) === 0)
        .filter((_, index) => index < 3)
    )
  }

  return result
}

module.exports = suggestedProducts

File: problems/sorting/1268-search-suggestions-system.py
"""#1268 Search Suggestions System
Sort once, then binary-search each prefix and return its first three matches.
Time: O(n log n * L + p log n + p^2) | Space: O(n + p), including the result.
Here L is the maximum product-name length and p is len(searchWord).
"""

from bisect import bisect_left


class Solution:
    def suggestedProducts(self, products, searchWord):
        products.sort()
        suggestions = []

        for end in range(1, len(searchWord) + 1):
            prefix = searchWord[:end]
            start = bisect_left(products, prefix)
            matches = []
            for product in products[start:start + 3]:
                if product.startswith(prefix):
                    matches.append(product)
                else:
                    break
            suggestions.append(matches)

        return suggestions
